### ISSUE TITLE
[Bug] verifyBracketsWellFormed not working as expected.

### ISSUE THREAD

**Issue by bh2smith** [gnosis/dex-liquidity-provision#217]
A few problems when trying to run bracket verification:

```
- Verify that masterCopy of brackets is the known masterCopy
Error: Returned values aren't valid, did it run Out of Gas?
    at getMasterCopy (/Users/admin/Projects/gnosis/dex-liquidity-provision/scripts/utils/internals.js:170:17)
```

```
- Verify proxy bytecode
AssertionError [ERR_ASSERTION]: Bad bytecode for bracket 0xfA4a18c2218945bC018BF94D093BCa66c88D3c40
```

Note that I am using the following safe setup on rinkeby:
```
masterSafe: 
    0xd9395aeE9141a3Efeb6d16057c8f67fBE296734c
brackets: 
    0xb947de73ADe9aBC6D57eb34B2CC2efd41f646636
    0xfA4a18c2218945bC018BF94D093BCa66c88D3c40
```

and that these brackets were created manually. It is possible that the second error is because the safes were created manually, but if this is the case I would expect the error message to be a bit more specific.

**Comment by bh2smith** [comment]
Can confirm, that I am also getting this error even for safes deployed with the project script:

```
- Verify proxy bytecode
AssertionError [ERR_ASSERTION]: Bad bytecode for bracket 0x725fd170B9a8c28caabB9b81Dd0dC80Fb3aa5Db6
    at Promise.all.bracketAddresses.map (/Users/admin/Projects/gnosis/dex-liquidity-provision/scripts/utils/verify_scripts.js:74:16)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Truffle v5.1.24 (core: 5.1.24)
```

**Steps to reproduce:**

```sh
truffle exec scripts/deploy_safes.js --masterSafe=$YOUR_SAFE --fleetSize 2 --network rinkeby
```
Copy the two new safe addresses and use them in the `--brackets` argument:

```sh
npx truffle exec scripts/complete_liquidity_provision.js --targetToken 1 --stableToken 4 --lowestLimit 200 --fleetSize 2 --highestLimit 216 --currentPrice 208 --masterSafe 0x9841644968e281445d59f3faae0ce96ca0f6ca13 --investmentTargetToken 0.1 --investmentStableToken 1 --network rinkeby --brackets=$NEW_SAFE1,$NEW_SAFE2
```

**Comment by bh2smith** [comment]
Can also confirm (at least on rinkeby) that the first error did not occur in this process.

**Comment by fedgiac** [comment]
The bracket address `0xb947de73ADe9aBC6D57eb34B2CC2efd41f646636` was created by the proxy factory at address `0x12302fE9c02ff50939BaAaaf415fc226C078613C`, while we use `0x76E2cFc1F5Fa8F6a5b3fC4c8F4788F0116861F9B` (safe version v1.1.1) for our brackets.

The first error is because the method to read the proxy template is not defined in the old version. The second error is because the bytecode of the safe does not match that of a proxy version v1.1.1.

Conclusion: the script was expected to fail on this input but the errors were unclear. The two checks should be swapped, so that the template is retrieved only if the contract is a verified safe. The error on failed bytecode verification should have a clearer error message, like "address x... is not a Gnosis Safe Proxy v1.1.1.

I can only run the scripts on Ganache, so I can't check what's causing the last failure.

**Comment by bh2smith** [comment]
I would expect the last failure to because of an old safe contract version.

**Comment by bh2smith** [comment]
So @fedgiac, if what you say is true, I should be able to upgrade the rinkeby safe contract from the interface, deploy two new safes and this should be fine?

**Comment by bh2smith** [comment]
I can confirm that I am getting this Bad bytecode error with safes deployed via the `deploy_safes` script:

```
AssertionError [ERR_ASSERTION]: Bad bytecode for bracket 0xf7Ec497D0fe288cE4D38e11027067Dfe7CB7F7AE
    at Promise.all.bracketAddresses.map (/Users/admin/Projects/gnosis/dex-liquidity-provision/scripts/utils/verify_scripts.js:74:16)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Truffle v5.1.25 (core: 5.1.25)
```

**Comment by bh2smith** [comment]
I have checked the contract code in the project dependencies:


https://unpkg.io/@gnosis.pm/safe-contracts@1.1.1-dev.2/contracts/proxies/GnosisSafeProxy.sol


and the code deployed at the bracket Address.

https://rinkeby.etherscan.io/address/0xddc07b251564496b9b09faa8f8fb770a97ec4912#code

Is is possible that the bytecode for rinkeby and mainnet are different and interfering?

**Comment by bh2smith** [comment]
After some awesome detective work by @josojo, it turns out this problem is rooted in the following solidity issue

https://github.com/ethereum/solidity/issues/7101

**Comment by fedgiac** [comment]
To my understanding, the difference between `deployedBytecode` and `bytecode` is that the former is the code of a contract as it's stored in the blockchain, while the latter is the code necessary to populate the blockchain with a new contract. So, for example, the code included in the constructor of a contract would appear in `bytecode` but not in `deployedBytecode`.

**Comment by bh2smith** [comment]
Sounds like you might have an idea how to reintroduce this verification?

**Comment by fedgiac** [comment]
Not really, I didn't look into the issue and I don't really know what might be causing this. Since it's relatively low priority I think it's fine to leave this issue open for a while.

**Comment by bh2smith** [comment]
Good suggestion by @rmeissner to compare against static bytecode-hash that we are willing to accept. This would certainly work. Will try to implement in the near future.

**Comment by fedgiac** [comment]
My summary of the meeting: our approach is bad because we compare the onchain bytecode to the bytecode we obtain from compiling the contract(s) ourselves, which can vary widely depending not only on the compiler version, but also on the machine it is compiled in because of [*to fill*](link here). 
This explains why everything is working with Ganache: the version we deploy comes from the proxy factory that we deployed ourself, which is compiled with the compiler we installed in our machine, while on mainnet we use an already existing proxy factory, which was compiled in a completely different way.

Suggested solutions:
1) keep a hardcoded list of all "supported" proxy bytecode versions.
2) ask the deployed proxy factory for the code of a proxy using [proxyCreationCode](https://github.com/gnosis/safe-contracts/blob/development/contracts/proxies/GnosisSafeProxyFactory.sol#L33).
3) 1+2: keep a hardcoded list of the bytecodes returned by the proxy factories we support.

I'd vote for 2, since we support only one proxy factory for each network.
